Merge backup-to-tarsnap and tarsnap into a single Ansible role (tarsnap)

taranjeet opened the pull request:
This PR removes `geerlinguy.java` submodule completely and merges `tarsnap` and `backup-to-tarsnap` role into a single role `tarsnap`. It also removes tarsnapper part(not in used) from `tarsnap` role.

taranjeet commented:
Some reference links
* [Remove git submodule](https://gist.github.com/myusuf3/7f645819ded92bda6677)

UmanShahzad commented on playbooks/roles/backup-swift-container/meta/main.yml:
```diff
@@ -1,7 +1,7 @@
 
 dependencies:
-  - tarsnap
+  - install_tarsnap
```
From what I see here, `install_tarsnap` is just a tag, right, and `tarsnap` is the role? AFAIK [from the documentation](https://docs.ansible.com/ansible/2.5/user_guide/playbooks_reuse_roles.html#role-dependencies), you cannot use tags here, you need to use actual role names.

taranjeet replied:
I just checked the role name will be `install` as per [this](https://github.com/open-craft/ansible-playbooks/pull/62/files#diff-7adcc8954ef0db37fa83c1a33ee2543b) file. I could have kept it to `tasks/tarsnap.yml` , but I guess on the task, @itsjeyd suggested [so](https://tasks.opencraft.com/browse/OC-4845?focusedCommentId=78914&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-78914).

What we can do is keep to tasks namely `tarsnap` and `backup_tarsnap`. @UmanShahzad is this right?

taranjeet replied:
@UmanShahzad : I messed this up. Thanks for pointing it out. Let me update the whole PR.
Btw, how do I test this?

UmanShahzad replied:
@taranjeet That's alright :) Testing locally will probably require you spin up a VM with Vagrant or similar, and configure your Ansible environment to run playbooks on it, which is nontrivial. It may be easiest to test on some stage server we have, e.g. RabbitMQ, but that may require significant access. I will let your reviewer @itsjeyd figure this part out with you.

UmanShahzad commented on playbooks/roles/tarsnap/tasks/main.yml:
```diff
@@ -1,7 +1,7 @@
 ---
 
-- include: tarsnap.yml tags=tarsnap
+- include: install.yml tags=install_tarsnap
```
I would prefer, if possible, to keep the tags be the same as the role names. That is the style we've been using elsewhere throughout the repo, e.g. the `filebeat` role will receive a `filebeat` tag so that we can just pass `--tag filebeat` if we want to run only that role. `install_` feels redundant unless there is a conflict somewhere giving undesirable behavior, and there really shouldn't be now that we're merging things.

UmanShahzad replied:
Nevermind, I see that this is just for a specific task file.

taranjeet commented:
@UmanShahzad : I hope things now makes sense. 
Here is what I have done. 

* `tarsnap` roles remains as it is. Its tasks are present in `roles/tarsnap/tasks/tarsnap.yml`. If particularly tasks of this role needs to be executed, one can simply use
```
{name: tarsnap, tags: tarsnap}
```

* `backup-to-tarsnap` role is now moved to `roles/tarsnap/tasks/configure_tarsnap.yml`. It can be called independenlty as 

```
{ name: tarsnap, tags: configure_tarsnap}
```

taranjeet commented:
@itsjeyd : Please review this and let me know how can I test this.

taranjeet commented:
@itsjeyd : Thanks for the review. I will go through it and update the PR as suggested.

taranjeet commented on playbooks/group_vars/all/public.yml:
```diff
@@ -7,7 +7,7 @@ ansible_python_interpreter: /usr/bin/python3
 
 tarsnap_version: 1.0.39
-tarsnap_cache: /var/tarsnap/cache
+tarsnap_cache_dir: /var/tarsnap/cache
```
@itsjeyd : Here is my [reason](https://tasks.opencraft.com/browse/OC-4845?focusedCommentId=78927&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-78927) for having two variable namely `tarsnap_cache_dir` and `TARSNAP_CACHE`. Please correct me if I am proceeding in the wrong direction.

itsjeyd replied:
@taranjeet Thanks, I had seen that comment :) But I'm not sure which part of it explains why we need to rename `tarsnap_cache` to `tarsnap_cache_dir` (and can't just keep `tarsnap_cache` and `TARSNAP_CACHE`)?

It seemed to me like you were mainly analyzing what the value of `tarsnap_cache` ultimately gets set to for the different invocations of the `tarsnap` role listed [here](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/backup-swift-container/meta/main.yml). Could you please clarify?

taranjeet replied:
@itsjeyd : I thought `tarsnap_cache` is an internal variable, and to avoid confusing, renaming it will be a better option. That way, one can easily get that `tarsnap_cache_dir` is an internal variable and whenever one needs to override that value, one can simply use `TARSNAP_CACHE` (like its happening at several places).

I will update this if this is not the right approach.

taranjeet commented on playbooks/roles/common-server/meta/main.yml:
```diff
@@ -14,8 +14,8 @@ dependencies:
-  - name: backup-to-tarsnap
-    tags: 'backup-to-tarsnap'
+  - name: tarsnap
+    tags: 'configure_tarsnap'
```
@itsjeyd : Does this means that whenever tarsnap role is requested, both the install and configure(which is actually the backup tarsnap part) should run?

I was thinking that sometimes there will be cases where we need to run `backup-to-tarnsap` role independently, hence I have kept the tag as `configure_tarsnap`.

itsjeyd replied:
@taranjeet 

> I was thinking that sometimes there will be cases where we need to run `backup-to-tarnsap` role independently, hence I have kept the tag as `configure_tarsnap`.

Did you find any concrete examples for this? I'd be surprised if you did -- the `backup-to-tarsnap` role included the `tarsnap` role as a dependency, so whenever the `backup-to-tarsnap` role was run as part of a playbook, the `tarsnap` role was run as well. (Which makes sense: The purpose of the `tarsnap` role was to install Tarsnap, and that's a prerequisite for the configuration steps that were encapsulated in the `backup-to-tarsnap` role.)

> Does this means that whenever tarsnap role is requested, both the install and configure(which is actually the backup tarsnap part) should run?

There might be cases where it's only necessary to run the installation portion of the new `tarsnap` role (an example for such a case is the backup pruner, cf. [`backup-pruner.yml`](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/backup-pruner.yml#L11-L12)). But based on the considerations above I don't think there are cases where we'd want to run the configuration part but skip the installation part.

---

In any case, this file was listing the `backup-to-tarsnap` role as a dependency, which means that the `common-server` role also depends on the (original) `tarsnap` role (i.e., the tasks that take care of installing Tarsnap). To properly reflect that, the tag should be `tarsnap` :)

taranjeet replied:
@itsjeyd : It took me some time but I am clear with it now. I have updated the PR as per this. :)

itsjeyd replied:
@taranjeet That's great, thanks :)

taranjeet commented:
@itsjeyd : Thanks a lot for such a detailed review. I will me go through it and update the code. :)

taranjeet commented on playbooks/roles/tarsnap/README.md:
```diff
@@ -17,17 +17,17 @@ Available variables are listed below, along with default values (see `defaults/m
+    tarsnap_cache_dir: "/usr/local/tarsnap-cache"
 
-The role-installed `.tarsnaprc` file will automatically exclude the `tarsnap_cache` folder and reference the `tarsnap_keyfile` for backups.  Your play or an out-of-band process is responsible for creating the keyfile and placing it at this location.  See the [standalone](#standalone) section.
+The role-installed `.tarsnaprc` file will automatically exclude the `tarsnap_cache_dir` folder and reference the `tarsnap_keyfile` for backups.  Your play or an out-of-band process is responsible for creating the keyfile and placing it at this location.
```
@itsjeyd : I agree. I have taken the liberty here to refactor README, which includes removing links which are not referred anywhere plus a bit of sentence restructuring. I am sure it will need more changes once the implementation part is over.

itsjeyd commented:
@taranjeet 

> I have incorporated the changes. Please review the PR. :)

For some reason GitHub submitted parts of my review before I had finished it. Not sure what happened there. Please have a look at the review I submitted just now (https://github.com/open-craft/ansible-playbooks/pull/62#pullrequestreview-139007170), it has some more stuff to follow up on.

I won't have time for another pass today, so feel free to take your time addressing the last round of comments.

taranjeet commented on playbooks/roles/tarsnap/tasks/main.yml:
```diff
@@ -1,7 +1,6 @@
-    - tarsnap
-  when: "{{ tarsnap_tarsnapper_conf != None}}"
+- include: install.yml tags=tarsnap
+- include: configure.yml tags=tarsnap
```
I get this. This was something which I totally missed it. Thanks for pointing it out. :)

taranjeet replied:
I have fixed this now. :)

itsjeyd replied:
@taranjeet Thanks! One more note about the include statements above: Can you put `tags` on a separate line below each include? For example:

```yaml
- include: install.yml
  tags:
    - tarsnap
```

The code below is using that style as well (instead of the `=` style), and not mixing styles will improve readability :)

taranjeet replied:
@itsjeyd : To conditionally run tasks from `configure.yml`, it should be `TARSNAP_KEY` rather than `TARSNAP_CACHE`, because `TARSNAP_CACHE` already has a value (=/var/cache/tarsnap). I have updated this.

taranjeet commented:
@itsjeyd : I have updated the PR. You can review :)

itsjeyd commented on playbooks/roles/tarsnap/tasks/main.yml:
```diff
@@ -4,3 +4,4 @@
 - include: configure.yml tags=tarsnap
   tarsnap_cache: "{{ TARSNAP_CACHE }}"
   tarsnap_keyfile: "{{ TARSNAP_KEY_REMOTE_LOCATION }}"
+  when: "{{ TARSNAP_CACHE != None}}"
```
@taranjeet Missing space after `None`.

itsjeyd commented on playbooks/roles/tarsnap/README.md:
```diff
@@ -2,8 +2,8 @@
+The role downloads sources for, verifies the gpg-encrypted sha signature, compiles, and installs [Tarsnap].
+**Batteries included**: cron job, shell wrapper, and [logrotate] policy.
 
 ## Requirements
```
@taranjeet This section contains a leftover broken link:

```
See [role variables](#Variables)
```

Could you just remove that part completely?

itsjeyd commented on playbooks/roles/tarsnap/defaults/main.yml:
```diff
@@ -1,12 +1,56 @@
 ---
 
+# Role Internal variables
```
@taranjeet Some formatting nits: Could you please

* insert a blank line below this one.
* indicate that this is a "section heading" by adding two more `#` chars. The result should look like this:
    ```
    ### Role-internal variables
    ```

itsjeyd commented on playbooks/roles/tarsnap/defaults/main.yml:
```diff
@@ -1,12 +1,56 @@
 #tarsnap_global_exclusions:
 #  - "*.deb"
+
+# Customizable variables
```
@taranjeet Same comment here: Please insert a blank line below this one, and lead with three `#` chars in total:

```
### Customizable variables
```

taranjeet commented:
@itsjeyd : All changes are incorporated. Thanks for the review. :)
Please have a look at it and if it looks neat, then this can be tested next.

itsjeyd commented on playbooks/roles/tarsnap/README.md:
```diff
@@ -13,49 +13,24 @@ The role downloads sources for, verifies the gpg-encrypted sha signature, compil
-    tarsnap_global_exclusions: # not defined by default
-
-List of string glob patterns that will be added to the `.tarsnaprc` file.  For an example list, see `personal/defaults.yml`
+For a complete list, see `defaults/main.yml`
```
@taranjeet This has not been addressed yet. You can hold off on it now, and just make the change as part of any follow-up that will come out of testing the changes.

taranjeet replied:
@itsjeyd : Sorry, I missed it somehow. I will take care of this in the changes which come when testing this PR. :)

taranjeet commented:
@itsjeyd : Thanks for the review. :)
I have updated the description of the PR and squashed commits to two(one for submodule and the other one for merging roles).
If these changes look good, shall I merge the PR myself?

taranjeet commented:
@itsjeyd : I looked into the last 5 merged PR's for this repo and inferred that its fine to merge the PR once it is approved and all the review comments are addressed. Here is the last review comment checklist
- [X] Update PR description
- [X] Squash commits into 2 commits(one for submodule and other one for merging roles).

Merging this PR. Thanks :)
